File: frontend/scratch/fix_translations.py
import os
import re
import json

src_dir = r"d:\allocento\frontend\src\app"
json_en = r"d:\allocento\frontend\src\assets\i18n\en.json"
json_hr = r"d:\allocento\frontend\src\assets\i18n\hr.json"

def process_ts_files():
    pattern_fallback = re.compile(r"(this\.toastService\.(?:success|error|warning|info))\s*\(\s*(?:translatedApiMsg\s*\|\|\s*)?this\.t\('([^']+)'\)\s*\|\|\s*'([^']+)'\s*\)")
    pattern_hardcoded = re.compile(r"(this\.toastService\.(?:success|error|warning|info))\s*\(\s*'([^']+)'\s*\)")
    
    extracted = {}

    for root, dirs, files in os.walk(src_dir):
        for file in files:
            if file.endswith(".ts"):
                filepath = os.path.join(root, file)
                with open(filepath, 'r', encoding='utf-8') as f:
                    content = f.read()

                new_content = content
                
                # Replace pattern 1: this.toastService.success(this.t('key') || 'Fallback')
                for match in pattern_fallback.finditer(content):
                    full_match = match.group(0)
                    method = match.group(1)
                    key = match.group(2)
                    fallback = match.group(3)
                    
                    extracted[key] = fallback
                    
                    if "translatedApiMsg" in full_match:
                        replacement = f"{method}(translatedApiMsg || this.t('{key}'))"
                    else:
                        replacement = f"{method}(this.t('{key}'))"
                    
                    new_content = new_content.replace(full_match, replacement)
                
                # Replace pattern 2: this.toastService.success('Hardcoded')
                for match in pattern_hardcoded.finditer(content):
                    full_match = match.group(0)
                    method = match.group(1)
                    hardcoded = match.group(2)
                    
                    # Generate a key based on the file name and a slug
                    base_name = os.path.basename(filepath).replace('.page.ts', '').replace('.component.ts', '')
                    slug = re.sub(r'[^a-zA-Z0-9]', '', hardcoded.title())[:15]
                    slug = slug[0].lower() + slug[1:]
                    key = f"{base_name}.{slug}"
                    
                    extracted[key] = hardcoded
                    
                    replacement = f"{method}(this.t('{key}'))"
                    new_content = new_content.replace(full_match, replacement)
                
                if new_content != content:
                    with open(filepath, 'w', encoding='utf-8') as f:
                        f.write(new_content)
                        
    return extracted

def fix_encoding(text):
    return text.replace('Ä\x8d', 'č').replace('Ä\x87', 'ć').replace('Å¡', 'š').replace('Å¾', 'ž').replace('Ä\x91', 'đ').replace('Ä\x8c', 'Č').replace('Ä\x86', 'Ć').replace('Å\xa0', 'Š').replace('Å½', 'Ž').replace('Ä\x90', 'Đ')

def update_json(filepath, extracted, is_hr=False):
    with open(filepath, 'r', encoding='utf-8') as f:
        data = json.load(f)
        
    for key_path, value in extracted.items():
        parts = key_path.split('.')
        current = data
        for i, part in enumerate(parts):
            if i == len(parts) - 1:
                # Always add if missing, or if it has bad characters
                if part not in current or "Ä" in current[part] or "Å" in current[part]:
                    if is_hr:
                        current[part] = fix_encoding(value)
                    else:
                        current[part] = value
            else:
                if part not in current:
                    current[part] = {}
                current = current[part]
                
    with open(filepath, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

if __name__ == "__main__":
    extracted = process_ts_files()
    print("Extracted strings:", len(extracted))
    for k, v in extracted.items():
        print(f"  {k}: {v}")
        
    update_json(json_en, extracted, is_hr=False)
    update_json(json_hr, extracted, is_hr=True)
    print("Updated JSON files.")

File: frontend/scratch/fix_translations.js
const fs = require('fs');
const path = require('path');

const srcDir = 'd:\\allocento\\frontend\\src\\app';
const jsonEn = 'd:\\allocento\\frontend\\src\\assets\\i18n\\en.json';
const jsonHr = 'd:\\allocento\\frontend\\src\\assets\\i18n\\hr.json';

function processTsFiles() {
    const patternFallback = /(this\.toastService\.(success|error|warning|info))\s*\(\s*(translatedApiMsg\s*\|\|\s*)?this\.t\('([^']+)'\)\s*\|\|\s*'([^']+)'\s*\)/g;
    const patternHardcoded = /(this\.toastService\.(success|error|warning|info))\s*\(\s*'([^']+)'\s*\)/g;
    
    const extracted = {};

    function walk(dir) {
        let results = [];
        const list = fs.readdirSync(dir);
        list.forEach(file => {
            file = path.join(dir, file);
            const stat = fs.statSync(file);
            if (stat && stat.isDirectory()) {
                results = results.concat(walk(file));
            } else if (file.endsWith('.ts')) {
                results.push(file);
            }
        });
        return results;
    }

    const files = walk(srcDir);
    
    files.forEach(filepath => {
        let content = fs.readFileSync(filepath, 'utf8');
        let newContent = content;
        
        // Replace pattern 1
        let match;
        while ((match = patternFallback.exec(content)) !== null) {
            const fullMatch = match[0];
            const method = match[1];
            const hasTranslatedApiMsg = match[3];
            const key = match[4];
            const fallback = match[5];
            
            extracted[key] = fallback;
            
            const replacement = hasTranslatedApiMsg 
                ? `${method}(translatedApiMsg || this.t('${key}'))`
                : `${method}(this.t('${key}'))`;
                
            newContent = newContent.replace(fullMatch, replacement);
        }
        
        // Replace pattern 2
        patternHardcoded.lastIndex = 0; // reset
        while ((match = patternHardcoded.exec(content)) !== null) {
            const fullMatch = match[0];
            const method = match[1];
            const hardcoded = match[3];
            
            let baseName = path.basename(filepath).replace('.page.ts', '').replace('.component.ts', '');
            let slug = hardcoded.replace(/[^a-zA-Z0-9]/g, ' ').split(' ').map(w => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase()).join('').substring(0, 15);
            slug = slug.charAt(0).toLowerCase() + slug.slice(1);
            if (!slug) slug = 'message';
            const key = `${baseName}.${slug}`;
            
            extracted[key] = hardcoded;
            
            const replacement = `${method}(this.t('${key}'))`;
            newContent = newContent.replace(fullMatch, replacement);
        }
        
        if (newContent !== content) {
            fs.writeFileSync(filepath, newContent, 'utf8');
        }
    });
                        
    return extracted;
}

function fixEncoding(text) {
    return text.replace(/Ä\x8d/g, 'č')
               .replace(/Ä\x87/g, 'ć')
               .replace(/Å¡/g, 'š')
               .replace(/Å¾/g, 'ž')
               .replace(/Ä\x91/g, 'đ')
               .replace(/Ä\x8c/g, 'Č')
               .replace(/Ä\x86/g, 'Ć')
               .replace(/Å\xA0/g, 'Š')
               .replace(/Å½/g, 'Ž')
               .replace(/Ä\x90/g, 'Đ');
}

function updateJson(filepath, extracted, isHr) {
    let data = JSON.parse(fs.readFileSync(filepath, 'utf8'));
        
    for (const [keyPath, value] of Object.entries(extracted)) {
        const parts = keyPath.split('.');
        let current = data;
        
        for (let i = 0; i < parts.length; i++) {
            const part = parts[i];
            if (i === parts.length - 1) {
                if (!current[part] || current[part] === "" || current[part].includes("Ä") || current[part].includes("Å")) {
                    current[part] = isHr ? fixEncoding(value) : value;
                }
            } else {
                if (!current[part]) {
                    current[part] = {};
                }
                current = current[part];
            }
        }
    }
                
    fs.writeFileSync(filepath, JSON.stringify(data, null, 2), 'utf8');
}

const extracted = processTsFiles();
console.log("Extracted strings:", Object.keys(extracted).length);
for (const [k, v] of Object.entries(extracted)) {
    console.log(`  ${k}: ${v}`);
}

updateJson(jsonEn, extracted, false);
updateJson(jsonHr, extracted, true);
console.log("Updated JSON files.");
